### ISSUE TITLE
Is there a Hook after Incrementing/Decrementing an Instance?

### ISSUE THREAD

**Issue by ms88privat** [sequelize/sequelize#4473]
"beforeUpdate"-Hook is not triggered if the Model Instance is incremented/decremented. Is there another way?

What I'm trying to archive is:
- updating a model instance multiple times in a very short period of time (avoid race conditions)
- only a part of the update is possible with increment/decrement (which should avoid race conditions?)
- the other value to update depends on the newly updated data (<b>how to do this without race conditions?</b>)

Using the build method is not possible because I have to determine the new value from the newly updated values, which i not have without race conditions.

``` javascript
// not possible because I don't know the actual values without race conditions?
modelInstance.build({}, {isNewRecord = false}).save(); 
```

So the possible solution I thought is, a Hook in combination with a race condition free increment/decrement method!? Any other possible sulutions?

Thank you 

EDIT:

I didn't realize that Increment returns a Promise and the new Data. So this solution should work without race conditions:

``` javascript
Counter.findById(...).then(counter =>
// this counter data is async and not necessary correct, right? 
// but it the doesn't matter here, because we are only incrementing now!
counter.increment({
   'count1': addValue1,
   'count2': addValue2
 }).then(counter => {
   // this counter data should be good to go, right?
    Counter.build({
        id: counter.id,
        greater: (counter.count1 >= counter.count2) ? 1 : 2
     },{isNewRecord: false})
     .save(['greater']);
 });
```

Any comments?

**Comment by mickhansen** [comment]
`increment/decrement` is built to avoid race conditions by doing an atomic update.
Because of the atomic update it's not possible to have any hooks.

AFAIR `increment` does not return the new data, but i could in postgres with  `RETURNING`

**Comment by ms88privat** [comment]
@mickhansen Thanks, I'm using postgres, so i will look deeper into `RETURNING`. 

But what is increment returning then? Only the actual data, which may or may not be the data after this specific increment, right?

**Comment by mickhansen** [comment]
Apparently the instance is returned: https://github.com/sequelize/sequelize/blob/master/lib/instance.js#L946

But it won't hold the updated value as the code stands at the moment.

**Comment by epicbagel** [comment]
@ms88privat looks like a recent change in #4253 means that the value is now in the returned object. As such, it should be possible to fire hooks. 

@mickhansen this may be a breaking change to people's code though (e.g. hooks will be triggered where they weren't before).

**Comment by mickhansen** [comment]
@epicbagel That change would not affect what hooks are called.

**Comment by epicbagel** [comment]
@mickhansen would you not trigger the before/afterUpdate hook as part of calling increment() though? Or, would you create an increment specific hook?

**Comment by mickhansen** [comment]
@epicbagel No, code: https://github.com/sequelize/sequelize/blob/master/lib/instance.js#L917 doesn't call any hooks at the moment.

**Comment by epicbagel** [comment]
@mickhansen ok, so we would implement a new beforeIncrement/afterIncrement hook? I could work on a PR if you think this would be the right approach.

**Comment by mickhansen** [comment]
@epicbagel It's the best cross dialect approach, for PG we could trigger afterUpdate on returning, but prefer to keep it generic as a first approach.

**Comment by ns3777k** [comment]
@epicbagel hey, any progress on this?

**Comment by exp10r3r** [comment]
Is this never going to be considered then?

**Comment by Strate** [comment]
It would great to have a hook to do some things after increment/decrement, for example send new value to socket

**Comment by pbrain19** [comment]
I implemented a simple solution to this problem. Make a new method under the class called incremementWithHook that encapsulates the increment and then whatever hook u need it to run. 

```

  async incrementWithHook(field: string): Promise<void> {
    await this.increment(field)
    await this.reload()
    await someHookHandler(this)
  }

```

Now instead of calling the hook directly you can use the newly created method.

**Comment by kevinlbatchelor** [comment]
Hate finding dead ends like this. what was wrong with after increment hook? lets reopen this.
